### ISSUE TITLE
Opam Files: Better Locations

### ISSUE THREAD

**Issue by rgrinberg** [ocaml/dune#8845]
At the moment, error messages when converting opam files to our lock directory leave much to be desired. The best we can do is point to the opam file that caused the error. It would be better if we could get the error to the concrete object (section, variable, filter, etc) that is causing the problem.

**Comment by Alizter** [comment]
Would this require modifications to the opam parser? Alternatively we could setup parsing opam files in our own dune_lang parser. There are a lot of similar features.

**Comment by rgrinberg** [comment]
Let's ask @rjbou what's the best way to proceed here. I would rather deviate from upstream as little as possible.

**Comment by rjbou** [comment]
Is it opam file format/parsing error messages?

**Comment by rgrinberg** [comment]
Nope, it's after parsing has been successful. It's at the stage of conversion from opam filters to dune's lock directory.
Here's an example from `opam_solver.ml`:

```ocaml
      List.map args ~f:(fun (simple_arg, _filter_opt) ->
        match simple_arg with
        | OpamTypes.CString s -> interpolate_opam_variables s
        | CIdent ident ->
          String_with_vars.make_pform
            Loc.none
            (Package_variable.pform_of_opam_ident (loc, OpamVariable.of_string ident)))
```

We would like that `loc` to come from the ident in `CIdent`, but it comes from the file at the moment.

**Comment by rjbou** [comment]
oh, code sample are only for current project

**Comment by Alizter** [comment]
Seems this is solved, would you agree @rgrinberg? I was testing it out and I am getting errors like:
```
  $ solve invalid-field
  Error: At
  $TESTCASE_ROOT/mock-opam-repository/packages/invalid-field/invalid-field.0.0.1/opam:2:10-2:11::
  Parse error
  [1]
```
However this is the location for the local repository, I have no idea what it will look like for the fetched one.

**Comment by rgrinberg** [comment]
Not really, it should show the snippet from the offending opam file.
